### PR TITLE
[sdpa] Allow prefill to use FA kernel with StaticCache

### PR DESCRIPTION
**Pull request by Cyrilvallez**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47094)

# What does this PR do?

As per the title. In general, we have no reason to constrain the prefill to have a mask with sdpa, as most usual situations will not need it. It's more efficient to let it dispatch to the FA kernel if we can. 

It also has the following benefits:
- much faster, with quadratic benefits as the prefill size increases (e.g. 10% faster for input size `(1, 512)` and max cache len `2048`, and a huge 260% faster for input size `(1, 8k)` and max cache length `16k`)
- aligns `_ignore_causal_mask_sdpa` to be the same for cuda and xpu, saving redundant code
- improve the previous xpu version, so that a model that returns `None` early via the `create_mask_for_generate` call will STILL return None quickly after reentering the mask creation function in the model (see issue https://github.com/huggingface/transformers/issues/46962 and PR https://github.com/huggingface/transformers/pull/47019), which should avoid headaches in the future

With a llama 8B with `32` layers, i.e. `32` attention calls during the prefill, an input size of `(1, 512)`, and a max cache length of `2048`, I measured the prefill (excluding cache lazy init) to take an average of `0.029` s with this PR, compared to `0.032` s on main -> 10% faster for small sizes.
For bigger sizes, the benefits are much greater: for size `(1, 8192)` and max cache length `16384`, prefill took `0.735` s on main vs `0.280` s on this PR -> 260% faster

Script used for benchmarking:
<details>

<summary>benchmark.py</summary>

```python
import torch
from transformers import AutoModelForCausalLM, StaticCache, GenerationConfig
import time

model_id = "meta-llama/Llama-3.1-8B-Instruct"
device = 0
prefill_size = 8192
max_len = 16384

model = AutoModelForCausalLM.from_pretrained(model_id, device_map=device)

ids = torch.randint(100, 200, (1, prefill_size), device=device)
generation_config = GenerationConfig()
model_kwargs = {
    "past_key_values": StaticCache(config=model.config, max_cache_len=max_len),
    "position_ids": torch.arange(prefill_size, device=device)[None, :],
    "attention_mask": torch.ones_like(ids)
}

# Lazy init the cache to avoid measuring cache allocation
with torch.no_grad():
    out = model._prefill(
        ids,
        generation_config=generation_config,
        model_kwargs=model_kwargs,
    )

N = 100
times = []
for _ in range(N):
    with torch.no_grad():
        # Reset cache to empty
        model_kwargs["past_key_values"].reset()
        t0 = time.time()
        out = model._prefill(
            ids,
            generation_config=generation_config,
            model_kwargs=model_kwargs,
        )
        dt = time.time() - t0
        times.append(dt)

print(f"Took {torch.mean(torch.tensor(times)):.3f} s in average")
```

</details>

cc @vasqu, @kaixuanliu for viz as well

### CODE REVIEW

File: src/transformers/integrations/sdpa_attention.py [hunk 1]
```diff
@@ -90,6 +93,16 @@ def sdpa_attention_forward(
             # Convert to boolean type, making sdpa to force call FlashAttentionScore to improve performance.
             attention_mask = torch.logical_not(attention_mask.bool()).to(query.device)
 
+    # This scenario can only happen during prefill with an empty StaticCache. Technically, since sdpa's `is_causal` mask alignment
+    # is upper-left, `is_causal=True` is enough to correctly compute the attention. However, sdpa will only dispatch to
+    # flash kernel if and only if q_length == kv_length, therefore it is more efficient to slice here and remove the masked tokens
+    # rather than to use the other available kernels for such a case.
+    # Note that we never compile prefill, and even if the user is doing it on its own, prefill and decode are 2 separate graphs
+    # anyway, so altering the shapes is fine here
```
Comment (ArthurZucker) [inline src/transformers/integrations/sdpa_attention.py:101]: we do compile it on prefix caching (the pr recently fromDavid) but yeah should be alright in most general cases

File: src/transformers/integrations/sdpa_attention.py [hunk 2]
```diff
@@ -90,6 +93,16 @@ def sdpa_attention_forward(
...
+    if is_causal and attention_mask is None and q_length > 1 and kv_length > q_length:
```
Comment (vasqu) [inline src/transformers/integrations/sdpa_attention.py:102]: 1. My main concern here is we inroduce a data dependent branching here. I'm unsure about the consequences here. I guess you will argue with attention mask in most cases not being None but still. Can you check with the new exporters PR from Ilyas if this doesn't break anything?
2. Do we not need to repad afterwards? 
3. Could this be moved to generate? I mean we technically can see whether we have an sdpa implementation + if we are in a static scenario. So should be easy to move?

File: src/transformers/masking_utils.py [hunk 3]
```diff
@@ -227,42 +227,9 @@ def maybe_pad_block_sequence_ids(
...
-    if q_length == 1:
-        # Single query token: skip only if no padding tokens present
-        return padding_mask.all()
-
-    # XPU-specific: check if query window is all True and rest is all False.
-    # This allows XPU to optimize the 1st token in static cache when the cache is empty.
-    return q_offset == 0 and padding_mask[:, :q_length].all() and not padding_mask[:, q_length:].any()
+def fast_all(tensor: torch.BoolTensor) -> torch.BoolTensor:
+    """Similar to `tensor.all()`, but uses an implementation with `tensor.sum()`, which is actually much faster."""
+    return tensor.sum() == tensor.numel()
```
Comment (vasqu) [inline src/transformers/masking_utils.py:232]: Oh wow how did you find that trick lol. But awesome!

File: src/transformers/masking_utils.py [hunk 4]
```diff
@@ -282,28 +249,29 @@ def _ignore_causal_mask_sdpa(
...
+        return False
+    # In this case, we need to add special patterns to the mask no matter what, so we cannot use any of the later skip conditions
+    if local_attention_size is not None and kv_length >= local_attention_size:
+        return False
+
+    # If `q_length == 1`, we then use `is_causal=False` in sdpa integration to mimic lower-right alignment. If `kv_length == q_length`,
+    # we use `is_causal=True` as upper-left alignment (torch's default) is the same as lower-right in this case. If we have padding,
+    # we need to add padding to the mask, so cannot be skipped
+    if (q_length == 1 or kv_length == q_length) and (padding_mask is None or fast_all(padding_mask)):
+        return True
```
Comment (vasqu) [inline src/transformers/masking_utils.py:269]: Is there not a misalignment for q_len == 1 in the comments? Should we or should we not skip in those instances?

Comment (vasqu) [reply src/transformers/masking_utils.py:269]: Definitely would double check whether SDPA handles this correctly (as XPU FA backend is different from GPU FA backend iirc)

**Review by vasqu — COMMENTED** [review]
Some quick comments from my side 🤗 

Imo, it might be nicer to move that prep outside to the static input generation as we could check attn implementation etc before, no? My goal is just to remove anything data dependent as best as we can

File: src/transformers/masking_utils.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/masking_utils.py:232]: Was profiling and noticed that `all` was super slow for some reason, so started experimenting other easy implems haha

File: src/transformers/masking_utils.py (hunk 4 above)
Comment (Cyrilvallez) [reply src/transformers/masking_utils.py:269]: No no it's handled correctly, and the only thing that made it work all this time, otherwise sdpa would have been broken for so long (just the comment was weird, I made it much clearer)

File: src/transformers/integrations/sdpa_attention.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/integrations/sdpa_attention.py:102]: 1. Technically it's not data-dependent as for torch compile, we only check shapes and None. For export, the mask CANNOT be None, as we have the strict condition `if not is_tracing`, so there's no worries at all there
2. No, no need to repad at all, the matrix output is the same (we remove the padding from both k and v)
3. We cannot move it to `generate` as we don't know there what shapes the k/v will take in general, and I don't think it would make much sense there (we would need to check so many cases of layer types etc)

**Comment by kaixuanliu** [comment]
Thanks @Cyrilvallez , I verified the correctness and benchmarked the performance on XPU as well:
| Settings | main | this PR | speedup
|-----|-----|-----|-----|
| seq=512, cache=2048   | 35.1 ms   | 32.7 ms   | ~7% |
| seq=2048, cache=8192   | 80.5ms   | 43.3 ms   | ~86%|
| seq=4096, cache=8192   | 148.3ms   | 78.5 ms   | ~89%|

basically align with what your data collected on cuda device.

**Comment by Cyrilvallez** [comment]
Thanks a lot for double checking perfs and correctness! 🤗

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/integrations/sdpa_attention.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/integrations/sdpa_attention.py:101]: In those cases, the `is_tracing` call will return True, so we will ALWAYS have a mask anyway (same as export), so it's all fine!
